autometa/common/exceptions.py

```diff
@@ -39,41 +38,118 @@ def __init__(self, value):
     You may file an issue with us at https://github.com/KwanLab/Autometa/issues/new
     """
 
+    def __init__(self, value, issue_request):
+        self.value = value
+        self.issue_request = issue_request
+
     def __str__(self):
-        return f"{self.value}\n\n{issue_request}"
+        return f"{self.value}\n\n{self.issue_request}"
 
 
 class KmerFormatError(AutometaException):
-    """KmerFormatError exception class."""
+    """
+    KmerFormatError exception class.
+
+    Parameters
+    ----------
+    AutometaException : class
+        Base class for other exceptions
+    """
 
     def __init__(self, fpath):
-        super(AutometaException, self).__init__(fpath)
         self.fpath = fpath
 
     def __str__(self):
+        """
+        Operator overloading to return the path to k-mers frequency tab-delimited table or
+        k-mers normalized tab-delimited table
+
+        Returns
+        -------
+        str
+            Path to k-mers frequency tab-delimited table or k-mers normalized tab-delimited table
+        """
         return (
             f'{self.fpath} does not contain a "contig" column. '
             "Ensure the k-mer matrix was properly generated."
         )
 
 
 class KmerEmbeddingError(AutometaException):
-    """KmerEmbeddingError exception class."""
+    """
+    KmerEmbeddingError exception class.
+
+    Parameters
+    ----------
+    AutometaException : class
+        Base class for other exceptions
+    """
 
     def __init__(self, value):
         self.value = value
 
     def __str__(self):
+        """
+        Operator overloading to return the text message written while raising the error,
+        rather than the message of __str__ by base exception
+
+        Returns
+        -------
+        str
+            Message written alongside raising the exception
+        """
         return self.value
 
 
 class BinningError(AutometaException):
-    """BinningError exception class."""
+    """
+    BinningError exception class.
+
+    Parameters
+    ----------
+    AutometaException : class
+        Base class for other exceptions
+    """
+
+    def __init__(self, value):
+        self.value = value
+
+    def __str__(self):
+        """
+        Operator overloading to return the text message written while raising the error,
+        rather than the message of __str__ by base exception
+
+        Returns
+        -------
+        str
+            Message written alongside raising the exception
+        """
+        return self.value
+
+
+class DatabaseOutOfSyncError(AutometaException):
```

Comment:
Since this is a handled exception, we probably do not need to emit to the user to file an issue request. Really they should just re-download their databases.

```suggestion
class DatabaseOutOfSyncError(Exception):
```

Reply:
From my reading, it is a good practice to have a base exception class when raising multiple custom made exceptions.
https://docs.python.org/3/tutorial/errors.html#user-defined-exceptions
- the `__str__` in `DatabaseOutOfSyncError` class would prevent from emitting to the user to file an issue request. as it is curently coded the following would happen:
```python
In [1]: from autometa.taxonomy.ncbi import NCBI                                                                                                                                                                                                                                     

In [2]: ncbi=NCBI("/home/the_bio_informatician/Autometa/autometa/databases/ncbi/")                                                                                                                                                                                                  
DatabaseWarning: /home/the_bio_informatician/Autometa/autometa/databases/ncbi/nr.gz needs to be formatted for diamond!

In [3]: ncbi.convert_taxid_dtype(5233333333333336)                                                                                                                                                                                                                                  
NCBI databases out of sync. 5233333333333336 not in found in any of the NCBI databases - nodes.dmp, names.dmp and merged.dmp.
---------------------------------------------------------------------------
DatabaseOutOfSyncError                    Traceback (most recent call last)
<ipython-input-3-9a7f87eb4f77> in <module>
----> 1 ncbi.convert_taxid_dtype(5233333333333336)

~/Autometa/autometa/taxonomy/ncbi.py in convert_taxid_dtype(self, taxid)
    448             err_message = f"NCBI databases out of sync. {taxid} not in found in any of the NCBI databases - nodes.dmp, names.dmp and merged.dmp."
    449             logger.warning(err_message)
--> 450             raise DatabaseOutOfSyncError(err_message)
    451         return taxid
    452 

DatabaseOutOfSyncError: NCBI databases out of sync. 5233333333333336 not in found in any of the NCBI databases - nodes.dmp, names.dmp and merged.dmp.

In [4]:  
```
This is because we are passing in `value`, which is nothing but the `err_message`. And as per the `__str__` specified in the `DatabaseOutOfSyncError` class that is emitted out rather than what is in its base class i.e. `AutometaException`
- Another reason to use base class is to maintain consistency as other custom exceptions like `KmerFormatError` are also being used at other places in the script.

Reply:
Alright, I think we keep the base `AutometaException` (maybe we rename this to `AutometaError`). 

So any time we do not want to emit the issue request we must overload the `__str__` operator? This seems like we would want to do this for every exception we write, which makes me think we should not have this `issue_request` message in the base class.

Reply:
But then what if the user experiences an error we didn't account for?

Reply:
This would mean we have not handled the error, therefore we would not have an `AutometaError` to wrap whatever `Exception` was raised. So the `issue_request` would not be emitted anyways. The solution for emitting the `issue_request` on any `Exception` is to wrap the Autometa entrypoints in try/except statements. If _any_ `Exception` is raised that _is not_ an `AutometaError` that was handled, then the `issue_request` will be emitted.